File: buyer-android/app/src/main/java/com/poultry/buyer/core/payment/PaymentManager.kt
package com.poultry.buyer.core.payment

import android.app.Activity
import com.poultry.buyer.BuildConfig
import com.poultry.buyer.domain.model.Payment
import com.razorpay.Checkout
import com.razorpay.PaymentData
import com.razorpay.PaymentResultWithDataListener
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Sealed class representing payment result states
 */
sealed class PaymentResult {
    data object Idle : PaymentResult()
    data object Processing : PaymentResult()
    data class Success(
        val razorpayPaymentId: String,
        val razorpayOrderId: String,
        val razorpaySignature: String
    ) : PaymentResult()
    data class Failure(
        val errorCode: Int,
        val errorDescription: String,
        val paymentData: PaymentData?
    ) : PaymentResult()
    data class ExternalWallet(
        val walletName: String
    ) : PaymentResult()
}

/**
 * Data class for payment options
 */
data class PaymentOptions(
    val orderId: String,
    val razorpayOrderId: String,
    val amount: Long, // Amount in paise (INR subunit)
    val currency: String = "INR",
    val name: String = "Poultry Platform",
    val description: String,
    val prefillEmail: String? = null,
    val prefillPhone: String? = null,
    val prefillName: String? = null,
    val notes: Map<String, String>? = null,
    val theme: PaymentTheme = PaymentTheme()
)

/**
 * Payment theme customization
 */
data class PaymentTheme(
    val color: String = "#4CAF50", // Green theme
    val backdropColor: String = "#ffffff"
)

/**
 * PaymentManager handles Razorpay payment integration.
 *
 * Usage:
 * 1. Call startPayment() with payment options from an Activity
 * 2. Activity must implement PaymentResultWithDataListener
 * 3. Handle results via the paymentResult StateFlow
 */
@Singleton
class PaymentManager @Inject constructor() {

    companion object {
        private const val TAG = "PaymentManager"

        /**
         * Call this during Application.onCreate() to preload Razorpay
         */
        fun preloadCheckout(context: android.content.Context) {
            Checkout.preload(context)
        }
    }

    private val _paymentResult = MutableStateFlow<PaymentResult>(PaymentResult.Idle)
    val paymentResult: StateFlow<PaymentResult> = _paymentResult.asStateFlow()

    private var currentPayment: Payment? = null

    /**
     * Initialize and start a Razorpay payment
     *
     * @param activity The activity context (must implement PaymentResultWithDataListener)
     * @param payment The payment object from API
     * @param buyerName Optional buyer name for prefill
     * @param buyerPhone Optional buyer phone for prefill
     * @param buyerEmail Optional buyer email for prefill
     */
    fun startPayment(
        activity: Activity,
        payment: Payment,
        buyerName: String? = null,
        buyerPhone: String? = null,
        buyerEmail: String? = null
    ) {
        currentPayment = payment
        _paymentResult.value = PaymentResult.Processing

        val checkout = Checkout()
        checkout.setKeyID(payment.razorpayKey ?: BuildConfig.RAZORPAY_KEY_ID)

        try {
            val options = JSONObject().apply {
                // Order details
                put("name", "Poultry Platform")
                put("description", "Order Payment")
                put("order_id", payment.razorpayOrderId ?: payment.gatewayOrderId)
                put("currency", payment.currency)

                // Amount in paise (multiply by 100)
                put("amount", (payment.amount * 100).toLong())

                // Prefill details
                val prefill = JSONObject()
                buyerEmail?.let { prefill.put("email", it) }
                buyerPhone?.let { prefill.put("contact", it) }
                buyerName?.let { prefill.put("name", it) }
                put("prefill", prefill)

                // Notes for reference
                val notes = JSONObject()
                notes.put("payment_id", payment.id)
                notes.put("order_id", payment.orderId)
                put("notes", notes)

                // Theme customization
                val theme = JSONObject()
                theme.put("color", "#4CAF50")
                theme.put("backdrop_color", "#ffffff")
                put("theme", theme)

                // Retry configuration
                val retry = JSONObject()
                retry.put("enabled", true)
                retry.put("max_count", 4)
                put("retry", retry)

                // Payment methods
                val modal = JSONObject()
                modal.put("confirm_close", true)
                modal.put("ondismiss", true)
                put("modal", modal)
            }

            checkout.open(activity, options)
        } catch (e: Exception) {
            _paymentResult.value = PaymentResult.Failure(
                errorCode = -1,
                errorDescription = e.message ?: "Failed to initialize payment",
                paymentData = null
            )
        }
    }

    /**
     * Start payment with custom options
     */
    fun startPaymentWithOptions(
        activity: Activity,
        options: PaymentOptions
    ) {
        _paymentResult.value = PaymentResult.Processing

        val checkout = Checkout()
        checkout.setKeyID(BuildConfig.RAZORPAY_KEY_ID)

        try {
            val jsonOptions = JSONObject().apply {
                put("name", options.name)
                put("description", options.description)
                put("order_id", options.razorpayOrderId)
                put("currency", options.currency)
                put("amount", options.amount)

                // Prefill
                val prefill = JSONObject()
                options.prefillEmail?.let { prefill.put("email", it) }
                options.prefillPhone?.let { prefill.put("contact", it) }
                options.prefillName?.let { prefill.put("name", it) }
                put("prefill", prefill)

                // Notes
                options.notes?.let { notes ->
                    val notesJson = JSONObject()
                    notes.forEach { (key, value) ->
                        notesJson.put(key, value)
                    }
                    put("notes", notesJson)
                }

                // Theme
                val theme = JSONObject()
                theme.put("color", options.theme.color)
                theme.put("backdrop_color", options.theme.backdropColor)
                put("theme", theme)

                // Retry
                val retry = JSONObject()
                retry.put("enabled", true)
                retry.put("max_count", 4)
                put("retry", retry)
            }

            checkout.open(activity, jsonOptions)
        } catch (e: Exception) {
            _paymentResult.value = PaymentResult.Failure(
                errorCode = -1,
                errorDescription = e.message ?: "Failed to initialize payment",
                paymentData = null
            )
        }
    }

    /**
     * Handle successful payment callback
     * Called by Activity implementing PaymentResultWithDataListener
     */
    fun onPaymentSuccess(paymentData: PaymentData) {
        onPaymentSuccess(
            razorpayPaymentId = paymentData.paymentId,
            razorpayOrderId = paymentData.orderId,
            razorpaySignature = paymentData.signature
        )
    }

    /**
     * Handle successful payment callback when only raw values are available
     */
    fun onPaymentSuccess(
        razorpayPaymentId: String?,
        razorpayOrderId: String?,
        razorpaySignature: String?
    ) {
        _paymentResult.value = PaymentResult.Success(
            razorpayPaymentId = razorpayPaymentId ?: "",
            razorpayOrderId = razorpayOrderId ?: "",
            razorpaySignature = razorpaySignature ?: ""
        )
    }

    /**
     * Handle payment failure callback
     * Called by Activity implementing PaymentResultWithDataListener
     */
    fun onPaymentError(code: Int, description: String?, paymentData: PaymentData?) {
        _paymentResult.value = PaymentResult.Failure(
            errorCode = code,
            errorDescription = description ?: getErrorMessage(code),
            paymentData = paymentData
        )
    }

    /**
     * Handle external wallet selection
     */
    fun onExternalWallet(walletName: String?) {
        _paymentResult.value = PaymentResult.ExternalWallet(
            walletName = walletName ?: "Unknown Wallet"
        )
    }

    /**
     * Reset payment state to idle
     */
    fun resetPaymentState() {
        _paymentResult.value = PaymentResult.Idle
        currentPayment = null
    }

    /**
     * Get current payment being processed
     */
    fun getCurrentPayment(): Payment? = currentPayment

    /**
     * Get human-readable error message for Razorpay error codes
     */
    private fun getErrorMessage(code: Int): String {
        return when (code) {
            Checkout.NETWORK_ERROR -> "Network error. Please check your connection and try again."
            Checkout.INVALID_OPTIONS -> "Invalid payment options. Please try again."
            Checkout.PAYMENT_CANCELED -> "Payment was cancelled."
            Checkout.TLS_ERROR -> "Security error. Please update your app or device."
            else -> "Payment failed. Please try again."
        }
    }
}

/**
 * Extension function to convert Payment amount to paise
 */
fun Payment.amountInPaise(): Long = (amount * 100).toLong()

/**
 * Data class for payment verification request
 */
data class PaymentVerificationRequest(
    val paymentId: String,
    val razorpayPaymentId: String,
    val razorpayOrderId: String,
    val razorpaySignature: String
)

File: buyer-ios/PoultryBuyer/Core/Payment/PaymentManager.swift
//
//  PaymentManager.swift
//  PoultryBuyer
//
//  Razorpay payment gateway integration wrapper
//

import Foundation
import UIKit

#if canImport(Razorpay)
import Razorpay
typealias RazorpayCheckoutProtocol = RazorpayCheckout
#else
/// Protocol for Razorpay SDK interaction
/// Implemented by the SDK when available.
protocol RazorpayCheckoutProtocol: AnyObject {
    func open(options: [String: Any], displayController: UIViewController)
    func setExternalWalletSelectionDelegate(_ delegate: Any)
}

/// Razorpay payment data delegate protocol
protocol RazorpayPaymentCompletionProtocol: AnyObject {
    func onPaymentSuccess(_ payment_id: String, andData response: [AnyHashable: Any]?)
    func onPaymentError(_ code: Int32, description str: String, andData response: [AnyHashable: Any]?)
}

/// External wallet selection delegate protocol
protocol ExternalWalletSelectionProtocol: AnyObject {
    func onExternalWalletSelected(_ walletName: String, withPaymentData paymentData: [AnyHashable: Any]?)
}
#endif

// MARK: - Payment Models

/// Represents a Razorpay payment order created on the server
struct PaymentOrder: Codable {
    let id: String
    let razorpayOrderId: String
    let orderId: String
    let amount: Int  // Amount in paise (smallest currency unit)
    let currency: String
    let status: PaymentOrderStatus
    let receipt: String?
    let notes: [String: String]?
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case razorpayOrderId
        case orderId
        case amount
        case currency
        case status
        case receipt
        case notes
        case createdAt
    }
}

enum PaymentOrderStatus: String, Codable {
    case created = "CREATED"
    case attempted = "ATTEMPTED"
    case paid = "PAID"
    case failed = "FAILED"
}

/// Request to create a payment order
struct CreatePaymentRequest: Codable {
    let orderId: String
    let amount: Int  // Amount in paise
    let currency: String
    let receipt: String?
    let notes: [String: String]?

    init(orderId: String, amount: Double, currency: String = "INR", receipt: String? = nil, notes: [String: String]? = nil) {
        self.orderId = orderId
        self.amount = Int(amount * 100)  // Convert to paise
        self.currency = currency
        self.receipt = receipt
        self.notes = notes
    }
}

/// Request to verify a payment
struct VerifyPaymentRequest: Codable {
    let razorpayOrderId: String
    let razorpayPaymentId: String
    let razorpaySignature: String
}

/// Response from payment verification
struct PaymentVerificationResponse: Codable {
    let verified: Bool
    let paymentId: String
    let orderId: String
    let status: String
    let message: String?
}

/// Payment status response
struct PaymentStatusResponse: Codable {
    let id: String
    let orderId: String
    let status: String
    let amount: Int
    let currency: String
    let razorpayPaymentId: String?
    let paidAt: Date?
    let failureReason: String?
}

// MARK: - Payment Result

/// Result of a Razorpay payment attempt
enum PaymentResult {
    case success(PaymentSuccessData)
    case failure(PaymentFailureData)
    case cancelled
}

struct PaymentSuccessData {
    let razorpayPaymentId: String
    let razorpayOrderId: String
    let razorpaySignature: String
}

struct PaymentFailureData {
    let code: Int
    let description: String
    let source: String?
    let step: String?
    let reason: String?
    let orderId: String?
    let paymentId: String?
}

// MARK: - Payment Delegate

protocol PaymentManagerDelegate: AnyObject {
    func paymentManager(_ manager: PaymentManager, didCompleteWith result: PaymentResult)
}

// MARK: - Razorpay Protocol (SDK Abstraction)

// MARK: - Payment Manager

/// Manages Razorpay payment integration
///
/// Usage:
/// 1. Initialize with delegate
/// 2. Call startPayment with order details
/// 3. Handle result via delegate callback
///
/// Note: Requires Razorpay iOS SDK to be installed via CocoaPods or SPM
/// Add to Podfile: pod 'razorpay-pod', '~> 1.3.0'
/// Or add SPM package: https://github.com/nicksth/razorpay-swift-package.git
@MainActor
class PaymentManager: NSObject {

    // MARK: - Properties

    static let shared = PaymentManager()

    weak var delegate: PaymentManagerDelegate?

    /// Razorpay SDK instance - set when SDK is configured
    /// In production, this would be: Razorpay.initWithKey(razorpayKey, andDelegate: self)
    private var razorpay: RazorpayCheckoutProtocol?

    /// Current payment order being processed
    private var currentPaymentOrder: PaymentOrder?

    /// Current order ID for tracking
    private var currentOrderId: String?

    /// Razorpay API key (loaded from configuration)
    private var razorpayKey: String {
        // In production, load from AppConfig or environment
        #if DEBUG
        return AppConfig.razorpayKeyId
        #else
        return AppConfig.razorpayKeyId
        #endif
    }

    // MARK: - Initialization

    private override init() {
        super.init()
        setupRazorpay()
    }

    private func setupRazorpay() {
        #if canImport(Razorpay)
        razorpay = RazorpayCheckout.initWithKey(razorpayKey, andDelegate: self)
        razorpay?.setExternalWalletSelectionDelegate(self)
        #else
        razorpay = nil
        #endif

        print("[PaymentManager] Initialized with key: \(razorpayKey.prefix(8))...")
    }

    // MARK: - Public Methods

    /// Start a payment flow for an order
    /// - Parameters:
    ///   - paymentOrder: The payment order created on the server
    ///   - customerInfo: Customer details for prefilling checkout
    ///   - presentingController: UIViewController to present checkout on
    func startPayment(
        paymentOrder: PaymentOrder,
        customerInfo: PaymentCustomerInfo,
        presentingController: UIViewController
    ) {
        currentPaymentOrder = paymentOrder
        currentOrderId = paymentOrder.orderId

        let options = buildPaymentOptions(
            paymentOrder: paymentOrder,
            customerInfo: customerInfo
        )

        print("[PaymentManager] Starting payment for order: \(paymentOrder.orderId)")
        print("[PaymentManager] Razorpay Order ID: \(paymentOrder.razorpayOrderId)")
        print("[PaymentManager] Amount: \(paymentOrder.amount) paise")

        // Open Razorpay checkout if SDK is available
        razorpay?.open(options: options, displayController: presentingController)

        // For development without SDK, simulate a successful payment after 2 seconds
        #if DEBUG
        if razorpay == nil {
            simulatePaymentForDevelopment(paymentOrder: paymentOrder)
        }
        #endif
    }

    /// Cancel any ongoing payment
    func cancelCurrentPayment() {
        print("[PaymentManager] Payment cancelled by user")
        currentPaymentOrder = nil
        currentOrderId = nil
        delegate?.paymentManager(self, didCompleteWith: .cancelled)
    }

    // MARK: - Private Methods

    private func buildPaymentOptions(
        paymentOrder: PaymentOrder,
        customerInfo: PaymentCustomerInfo
    ) -> [String: Any] {
        var options: [String: Any] = [
            "key": razorpayKey,
            "amount": paymentOrder.amount,
            "currency": paymentOrder.currency,
            "order_id": paymentOrder.razorpayOrderId,
            "name": AppConfig.merchantName,
            "description": "Order #\(paymentOrder.receipt ?? paymentOrder.orderId)",
            "timeout": 300,  // 5 minutes timeout
            "send_sms_hash": true,
            "remember_customer": true
        ]

        // Prefill customer information
        var prefill: [String: String] = [:]
        if let name = customerInfo.name {
            prefill["name"] = name
        }
        if let email = customerInfo.email {
            prefill["email"] = email
        }
        if let phone = customerInfo.phone {
            prefill["contact"] = phone
        }
        options["prefill"] = prefill

        // Theme customization
        options["theme"] = [
            "color": "#FF6B35"  // App primary color
        ]

        // Notes for tracking
        var notes: [String: String] = [
            "app_order_id": paymentOrder.orderId
        ]
        if let additionalNotes = paymentOrder.notes {
            notes.merge(additionalNotes) { _, new in new }
        }
        options["notes"] = notes

        // Retry configuration
        options["retry"] = [
            "enabled": true,
            "max_count": 3
        ]

        // Payment method preferences
        options["config"] = [
            "display": [
                "blocks": [
                    "banks": [
                        "name": "Pay via UPI/Bank",
                        "instruments": [
                            ["method": "upi"],
                            ["method": "netbanking"]
                        ]
                    ],
                    "other": [
                        "name": "Other Payment Methods",
                        "instruments": [
                            ["method": "card"],
                            ["method": "wallet"]
                        ]
                    ]
                ],
                "sequence": ["block.banks", "block.other"],
                "preferences": [
                    "show_default_blocks": true
                ]
            ]
        ]

        return options
    }

    #if DEBUG
    /// Simulate payment for development testing without SDK
    private func simulatePaymentForDevelopment(paymentOrder: PaymentOrder) {
        print("[PaymentManager] DEVELOPMENT MODE: Simulating payment...")

        // Simulate payment processing delay
        Task {
            try? await Task.sleep(nanoseconds: 2_000_000_000)  // 2 seconds

            // Simulate successful payment
            let successData = PaymentSuccessData(
                razorpayPaymentId: "pay_\(UUID().uuidString.prefix(14))",
                razorpayOrderId: paymentOrder.razorpayOrderId,
                razorpaySignature: "simulated_signature_\(UUID().uuidString)"
            )

            print("[PaymentManager] DEVELOPMENT MODE: Payment simulated successfully")
            self.delegate?.paymentManager(self, didCompleteWith: .success(successData))
        }
    }
    #endif
}

// MARK: - Razorpay SDK Delegate Implementation

#if canImport(Razorpay)
extension PaymentManager: RazorpayPaymentCompletionProtocol {

    func onPaymentSuccess(_ payment_id: String, andData response: [AnyHashable: Any]?) {
        print("[PaymentManager] Payment succeeded: \(payment_id)")

        guard let response = response,
              let razorpayOrderId = response["razorpay_order_id"] as? String,
              let razorpaySignature = response["razorpay_signature"] as? String else {
            print("[PaymentManager] Error: Missing payment response data")
            let failureData = PaymentFailureData(
                code: -1,
                description: "Invalid payment response",
                source: nil,
                step: nil,
                reason: "Missing signature or order ID",
                orderId: currentOrderId,
                paymentId: payment_id
            )
            delegate?.paymentManager(self, didCompleteWith: .failure(failureData))
            return
        }

        let successData = PaymentSuccessData(
            razorpayPaymentId: payment_id,
            razorpayOrderId: razorpayOrderId,
            razorpaySignature: razorpaySignature
        )

        currentPaymentOrder = nil
        currentOrderId = nil

        delegate?.paymentManager(self, didCompleteWith: .success(successData))
    }

    func onPaymentError(_ code: Int32, description str: String, andData response: [AnyHashable: Any]?) {
        print("[PaymentManager] Payment failed: \(code) - \(str)")

        let failureData = PaymentFailureData(
            code: Int(code),
            description: str,
            source: response?["source"] as? String,
            step: response?["step"] as? String,
            reason: response?["reason"] as? String,
            orderId: response?["order_id"] as? String ?? currentOrderId,
            paymentId: response?["payment_id"] as? String
        )

        currentPaymentOrder = nil
        currentOrderId = nil

        delegate?.paymentManager(self, didCompleteWith: .failure(failureData))
    }
}

extension PaymentManager: ExternalWalletSelectionProtocol {
    func onExternalWalletSelected(_ walletName: String, withPaymentData paymentData: [AnyHashable: Any]?) {
        print("[PaymentManager] External wallet selected: \(walletName)")
    }
}
#endif

// MARK: - Customer Info

struct PaymentCustomerInfo {
    let name: String?
    let email: String?
    let phone: String?

    init(name: String? = nil, email: String? = nil, phone: String? = nil) {
        self.name = name
        self.email = email
        self.phone = phone
    }
}

// MARK: - App Config Extension for Payment

extension AppConfig {

    /// Razorpay Key ID
    static var razorpayKeyId: String {
        switch environment {
        case .development:
            // Test key for development
            return "rzp_test_XXXXXXXXXX"
        case .staging:
            return "rzp_test_XXXXXXXXXX"
        case .production:
            // Live key for production
            return "rzp_live_XXXXXXXXXX"
        }
    }

    /// Merchant name displayed in checkout
    static var merchantName: String {
        return "Poultry Platform"
    }
}
